### DISCUSSION TITLE
How to modify request scheme （https - > http）

### DISCUSSION THREAD

**Discussion by fairytaildhk** [mitmproxy/mitmproxy#4906] [Q&A] [answered]
I pull the project, I want to use mitmproxy with my mock service. Then I have write some code in “http.py”
The request is https, but my mock service host is http, In “class Request(Message)” this class I modify the scheme = 'http' but not take effect.
can you help me, thanks very much

the code:
class Request(Message):
    """
    An HTTP request.
    """
    data: RequestData

    def __init__(
        self,
        host: str,
        port: int,
        method: bytes,
        scheme: bytes,
        authority: bytes,
        path: bytes,
        http_version: bytes,
        headers: Union[Headers, Tuple[Tuple[bytes, bytes], ...]],
        content: Optional[bytes],
        trailers: Union[Headers, Tuple[Tuple[bytes, bytes], ...], None],
        timestamp_start: float,
        timestamp_end: Optional[float],
    ):

        # 是否走mock逻辑
        if method.decode("utf-8") != 'OPTIONS' and hulk_api.proxy_query_url_info(path.decode("utf-8"), headers) > 0:
            print(method.decode("utf-8"))
            print(('命中mock，接口：' + str(path)).center(100, '='))
            host = config.HULK_HOST
            scheme = 'http'
            port = 80
            headers.__delitem__('Host')
            headers.insert(0, b'Host', host)

**Comment by mhils** [comment, marked as answer]
Hi @fairytaildhk!

I'm not sure if I understand what your problem is. If you want to redirect requests to your mock service, take a look at the following resources:

- https://github.com/mitmproxy/mitmproxy/blob/main/examples/addons/http-redirect-requests.py
- https://github.com/mitmproxy/mitmproxy/blob/main/examples/addons/http-reply-from-proxy.py
- https://docs.mitmproxy.org/dev/overview-features/#map-remote

If you are redirecting from HTTPS to HTTP, make sure to update both `scheme` and `port`. Alternatively, you can update `flow.request.url`, which sets both of them.

**Reply by fairytaildhk** [reply]
thanks very much
Now I Use the http-redirect-requests.py update the  `flow.request.scheme`, it take effect
This frame is very good, I will continue study it, Do you have any suggest for me 😄
